commit 4e56cde15f7d68cf86ff8efff8504497de152475
Author: Jouni Malinen <quic_jouni@quicinc.com>
Date:   Fri Jul 31 21:38:30 2020 +0300

    mac80211: Handle special status codes in SAE commit
    
    SAE authentication has been extended with H2E (IEEE 802.11 REVmd) and PK
    (WFA) options. Those extensions use special status code values in the
    SAE commit messages (Authentication frame with transaction sequence
    number 1) to identify which extension is in use. mac80211 was
    interpreting those new values as the AP denying authentication and that
    resulted in failure to complete SAE authentication in some cases.
    
    Fix this by adding exceptions for the new status code values 126 and
    127.
    
    Signed-off-by: Jouni Malinen <jouni@codeaurora.org>
    Link: https://lore.kernel.org/r/20200731183830.18735-1-jouni@codeaurora.org
    Signed-off-by: Johannes Berg <johannes.berg@intel.com>

diff --git a/net/mac80211/mlme.c b/net/mac80211/mlme.c
index 839d0367446c..8b7ca8ddfe20 100644
--- a/net/mac80211/mlme.c
+++ b/net/mac80211/mlme.c
@@ -2988,7 +2988,10 @@ static void ieee80211_rx_mgmt_auth(struct ieee80211_sub_if_data *sdata,
 		cfg80211_rx_mlme_mgmt(sdata->dev, (u8 *)mgmt, len);
 
 		if (auth_alg == WLAN_AUTH_SAE &&
-		    status_code == WLAN_STATUS_ANTI_CLOG_REQUIRED)
+		    (status_code == WLAN_STATUS_ANTI_CLOG_REQUIRED ||
+		     (auth_transaction == 1 &&
+		      (status_code == WLAN_STATUS_SAE_HASH_TO_ELEMENT ||
+		       status_code == WLAN_STATUS_SAE_PK))))
 			return;
 
 		sdata_info(sdata, "%pM denied authentication (status %d)\n",